File: backend/app/api/v1/alerts.py
from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.repositories.alert_repo import alert_event_repo, alert_rule_repo
from app.repositories.agent_repo import agent_repo
from app.schemas.alert import AlertEventOut, AlertRuleCreate, AlertRuleOut

router = APIRouter(tags=["alerts"])


# ── Rules ──────────────────────────────────────────────────────────────────────

@router.post("/alerts/rules", response_model=AlertRuleOut, status_code=201)
def create_rule(
    payload: AlertRuleCreate,
    db: Session = Depends(get_db),
):
    """Create a new alert rule for an agent metric."""
    if not agent_repo.get_by_id(db, payload.agent_id):
        raise HTTPException(status_code=404, detail="Agent not found")
    rule = alert_rule_repo.create(db, **payload.model_dump())
    return rule


@router.get("/alerts/rules", response_model=List[AlertRuleOut])
def list_rules(
    agent_id: Optional[str] = None,
    db: Session = Depends(get_db),
):
    """List all alert rules, optionally filtered by agent."""
    if agent_id:
        return alert_rule_repo.list_for_agent(db, agent_id)
    return alert_rule_repo.list_all(db)


@router.patch("/alerts/rules/{rule_id}", response_model=AlertRuleOut)
def toggle_rule(
    rule_id: int,
    enabled: bool,
    db: Session = Depends(get_db),
):
    """Enable or disable a rule."""
    rule = alert_rule_repo.update_enabled(db, rule_id, enabled)
    if not rule:
        raise HTTPException(status_code=404, detail="Rule not found")
    return rule


@router.delete("/alerts/rules/{rule_id}", status_code=204)
def delete_rule(
    rule_id: int,
    db: Session = Depends(get_db),
):
    """Delete an alert rule."""
    if not alert_rule_repo.delete(db, rule_id):
        raise HTTPException(status_code=404, detail="Rule not found")


# ── Events (incident log) ──────────────────────────────────────────────────────

@router.get("/alerts/events", response_model=List[AlertEventOut])
def list_events(
    agent_id: Optional[str] = None,
    limit: int = 100,
    db: Session = Depends(get_db),
):
    """Retrieve the most recent alert events (incident log)."""
    return alert_event_repo.list_recent(db, limit=min(limit, 500), agent_id=agent_id)

File: dashboard/src/api/alerts.js
import { api } from './client.js'

export const alertsApi = {
  /** GET /api/v1/alerts/rules?agent_id=... */
  listRules: (agentId) => {
    const qs = agentId ? `?agent_id=${agentId}` : ''
    return api.get(`/api/v1/alerts/rules${qs}`)
  },

  /** POST /api/v1/alerts/rules */
  createRule: (payload) => api.post('/api/v1/alerts/rules', payload),

  /** PATCH /api/v1/alerts/rules/{id}?enabled=true|false */
  toggleRule: (ruleId, enabled) =>
    api.patch(`/api/v1/alerts/rules/${ruleId}?enabled=${enabled}`, {}),

  /** DELETE /api/v1/alerts/rules/{id} */
  deleteRule: (ruleId) => api.delete(`/api/v1/alerts/rules/${ruleId}`),

  /** GET /api/v1/alerts/events?agent_id=...&limit=100 */
  listEvents: (agentId, limit = 100) => {
    const params = new URLSearchParams({ limit })
    if (agentId) params.set('agent_id', agentId)
    return api.get(`/api/v1/alerts/events?${params}`)
  },
}
